Refactor switch for vesync

cdnninja opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Refactor of switch sensor to support it offering more switches in the future, such as screen on off.   This one will wait on the update coordinator merge first then update to reflect those changes.   Creating a draft PR for early feedback. 

I do not own a vesync switch or outlet so I would love to have someone test this on real hardware. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

cdnninja committed:
Refactor switch sensor so we can use it for more.

cdnninja commented:
@iprak FYI.  As two side notes you should add yourself as a code owner on this integration, your support has been incredible.     Would love to connect more on the HA discord too.

cdnninja committed:
Merge branch 'dev' into sensorRefactor

cdnninja commented:
I am thinking this PR should probably split the removal of the VeSyncDevice entity to a different PR. I don't see value in that class.  Just complicates things.   Thoughts?

cdnninja committed:
Adjust order so all platforms match between PRs

cdnninja committed:
Correct type

cdnninja committed:
Attempt to fix mypy error

cdnninja committed:
Correct mypy erros.

cdnninja committed:
Correct Errors

cdnninja committed:
Add callbacks

cdnninja committed:
Merge branch 'dev' into sensorRefactor

cdnninja committed:
Add coordinator as per other PR

cdnninja commented:
My one thought is I am not certain the names of the entities are unchanged for the two existing. I also can't test that hardware as I don't own it.

iprak commented:
> I am thinking this PR should probably split the removal of the VeSyncDevice entity to a different PR. I don't see value in that class. Just complicates things. Thoughts?

Sure, it definitely helps to keep PR small.

cdnninja commented:
> > I am thinking this PR should probably split the removal of the VeSyncDevice entity to a different PR. I don't see value in that class. Just complicates things. Thoughts?
> 
> Sure, it definitely helps to keep PR small.

Other portion has been split to https://github.com/home-assistant/core/pull/134499.

cdnninja committed:
Merge branch 'dev' into sensorRefactor

cdnninja committed:
Merge issue

cdnninja committed:
ruff

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -1,29 +1,54 @@
+
+
+SENSOR_DESCRIPTIONS: Final[tuple[VeSyncSwitchEntityDescription, ...]] = (
+    VeSyncSwitchEntityDescription(
+        key="device_status",
+        translation_key="on",
+        is_on=lambda device: device.device_status == "on",
+    ),
```
Should we just use the power device class? We should either set the name to `None` (causing the entity to follow the device name, so `switch.vesync_device`) or if there already is another main feature of the device (say a climate entity or humidifier), we should use the device class translation for the POWER device class

cdnninja replied:
Any harm in us not filtering if say it is a fan or humidifer?  name would be switch.vesync_device and allow a toggle for power.  This keeps the code simple as I don't need to filter if one of those.

iprak replied:
Duplicate entity? Humidifier and fan both derive from ToggleEntity and expose the same actions.

cdnninja replied:
I will test this in conjunction with humidifier PR once that one is merged.   Lets plan to merge this after that one.

iprak replied:
Sorry but why do we need this entity?

The `device_status` in [pyvesync ](https://github.com/webdjoe/pyvesync/blob/master/src/pyvesync/vesyncfan.py) seems to reflect the on/off state which fan/humidifier entities will report as well via `is_on`. They also have turn_on/turn_off to change the status. I did not check but I think the same should be true for light entities.

cdnninja replied:
They used to sell outlets and switches.  As per: https://www.home-assistant.io/integrations/vesync/ so this would show on off for the light switch or outlet.  I am thinking I should add a filter ability so it can be removed for select situations.    In the future this switch platform can offer other items though, which is the driver for the refactor.

iprak replied:
Switch and outlet would need that but I don't think they should be needed for fan/humidifiers.

My humidifier appears as following with off state.
![image](https://github.com/user-attachments/assets/9cfc228f-d9a7-40a4-bb3c-b2491b404171)
![image](https://github.com/user-attachments/assets/8508a9a3-6609-4060-ba6d-0e0f176a282f)

cdnninja replied:
I will add an additional check so we can adjust for that.  I won't get to this right a way since I think humidifier merge and binary_sensor take priority.

cdnninja replied:
Added the checks to ensure this is just outlet/switch.

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -46,57 +71,46 @@ def _setup_entities(
+            for description in SENSOR_DESCRIPTIONS
+            if rgetattr(dev, description.key) is not None
+        ),
+        update_before_add=True,
```
We use a coordinator so we already have a state so we don't need to request an update before that

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -46,57 +71,46 @@ def _setup_entities(
 
-    _attr_name = None
+class VeSyncSwitchEntity(SwitchEntity, VeSyncBaseEntity):
+    """VeSync sensor class."""
```
stale docstring

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -46,57 +71,46 @@ def _setup_entities(
-        return self.device.device_status == "on"
+    def is_on(self) -> bool | None:
+        """Return the entity value to represent the entity state."""
+        if self.entity_description.is_on is not None:
+            return self.entity_description.is_on(self.device)
+        return None
```
this is never none

cdnninja committed:
Fedback

cdnninja committed:
Merge branch 'dev' into sensorRefactor

cdnninja committed:
Ruff correction

cdnninja committed:
Add exists FN

cdnninja committed:
Corrections to pass tests

cdnninja committed:
Remove dimmable switch from scope to pass tests.

cdnninja committed:
Remove unused code to increase test coverage

cdnninja committed:
switch to async updates and refresh

cdnninja committed:
Revert last change

joostlek commented on homeassistant/components/vesync/common.py:
```diff
@@ -32,3 +34,15 @@ def is_humidifier(device: VeSyncBaseDevice) -> bool:
+
+
+def is_outlet(device: VeSyncBaseDevice) -> bool:
+    """Check if the device represents a fan."""
+
+    return isinstance(device, VeSyncOutlet)
+
+
+def is_wall_switch(device: VeSyncBaseDevice) -> bool:
+    """Check if the device represents a fan."""
```
```suggestion
    """Check if the device represents an outlet."""

    return isinstance(device, VeSyncOutlet)


def is_wall_switch(device: VeSyncBaseDevice) -> bool:
    """Check if the device represents a wall switch."""
```

joostlek commented on homeassistant/components/vesync/strings.json:
```diff
@@ -43,6 +43,11 @@
         "name": "Current voltage"
       }
     },
+    "switch": {
+      "on": {
+        "name": "On"
+      }
+    },
```
This is unused right?

cdnninja replied:
Removed.

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -1,29 +1,57 @@
+    VeSyncSwitchEntityDescription(
+        key="device_status",
+        is_on=lambda device: device.device_status == "on",
+        # Other types of wall switches support dimming.  Those use light.py platform.
+        exists_fn=lambda device: is_wall_switch(device) or is_outlet(device),
```
Okay, but how come only these 2 are switches? As in, `is_wall_switch` doesnt sound to incorporate dimming

cdnninja replied:
I thought I had handled this different maybe lost in a merge - so here is the challenge.  The old lookup table takes two model of switches and calls them a dimmer switch.  Goal is to not need that lookup table for models in the future. 

https://github.com/cdnninja/core/blob/09ae388f4e947f508cedda70b806a839326ea4af/homeassistant/components/vesync/const.py#L51-L52

These devices in the old method didn't create a switch entity just a light entity.    Currently with my code it would also get a switch.    Should we leave it like that so if dimmable switch or outlet you still get a switch toggle?      This is the easier way, other option is adjust the exists to exclude the type of VeSyncDimmerSwitch which is a type of switch within the library.  

@iprak  Thoughts?

iprak replied:
I don't think we need to provided a separate switch for dimmer. [LightEntity extends from ToggleEntity](https://github.com/home-assistant/core/blob/2bedb2cadbfd590c9c94b5a93793c83ce286d55e/homeassistant/components/light/__init__.py#L872) and so it provided on/off capability natively.

This is what I see for a dimmer switch
![image](https://github.com/user-attachments/assets/9311d1f5-aa26-4c46-a578-a63e78adbc8e)

cdnninja replied:
Went back to fix this and realized I was correct up front.   In pyvesync vesyncswitch has two child classes, VeSyncWallSwitch and VeSyncDimmerSwitch.   The above filters to just wall switch so dimmer isn't used.   As such this won't show for dimmers.   No changes required.

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +73,45 @@ def _setup_entities(
-    async_add_entities(entities, update_before_add=True)
-
+    """Check if device is online and add entity."""
+    async_add_entities(
+        (
+            VeSyncSwitchEntity(dev, description, coordinator)
+            for dev in devices
+            for description in SENSOR_DESCRIPTIONS
+            if description.exists_fn(dev)
+        ),
+    )
```
```suggestion
    async_add_entities(
        VeSyncSwitchEntity(dev, description, coordinator)
        for dev in devices
        for description in SENSOR_DESCRIPTIONS
        if description.exists_fn(dev)
    )
```

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +73,45 @@ def _setup_entities(
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__(device, coordinator)
+        self.entity_description: VeSyncSwitchEntityDescription = description
```
overwrite the type outside of the constructor instead of here, it usually looks a bit cleaner

cdnninja replied:
Adjusted.

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +73,45 @@ def _setup_entities(
+        """Initialize the sensor."""
+        super().__init__(device, coordinator)
+        self.entity_description: VeSyncSwitchEntityDescription = description
+        if isinstance(self.device, VeSyncOutlet):
+            self._attr_device_class = SwitchDeviceClass.OUTLET
+        if isinstance(self.device, VeSyncSwitch):
+            self._attr_device_class = SwitchDeviceClass.SWITCH
```
You have a helper for this

cdnninja replied:
Done.

joostlek commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +73,45 @@ def _setup_entities(
-        """Turn the device off."""
+        """Turn the entity off."""
         self.device.turn_off()
+        self.schedule_update_ha_state()
```
Why do we do that with the state?

cdnninja replied:
Without it HA toggles it back the other way until next scan interval which gives it a poor user experience as it appears to have not worked.

cdnninja committed:
Update homeassistant/components/vesync/common.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

cdnninja committed:
Update homeassistant/components/vesync/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

cdnninja committed:
Feedback Items

cdnninja committed:
Improved comments.

iprak commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +71,45 @@ def _setup_entities(
+        self.entity_description = description
+        if is_outlet(self.device):
+            self._attr_device_class = SwitchDeviceClass.OUTLET
+        if is_wall_switch(self.device):
```
We can combine these two to elif

cdnninja replied:
Fixed.

iprak commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +71,45 @@ def _setup_entities(
-    def turn_on(self, **kwargs: Any) -> None:
-        """Turn the device on."""
-        self.device.turn_on()
+    def __init__(
```
We need to define `_attr_unique_id` otherwise adding another switch entity fails.

Something like this will work - `self._attr_unique_id = f"{super().unique_id}-{description.key}"`

cdnninja replied:
For the two devices exposed today if we change this it will be a breaking change.    However I could add a filter to only do the above unique ID on ones other than the outlet and switch on we offer today.  It would set us up for future. 

Thoughts on that approach?

iprak replied:
I suppose that will work. @joostlek might have reviewed code which ran into similar shortcoming and if the _attr_unique_id was defined as a breaking change.

joostlek replied:
Oh that is a good catch. If that is the case we should migrate first. I don't feel like merging this as it will create entities with not unique unique ids. So we should have a preliminary PR where we migrate all the unique ids of the switches to something else and then we can set up new ones

cdnninja replied:
Okay,  today this code doesn't conflict or impact unique ID, however since trying to setup for growth makes sense we fix this now.   I will update this code to have the unique ID as mentioned above and chat with iprak around getting a PR for that migration that will merge first.

iprak commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +71,45 @@ def _setup_entities(
-        """Turn the device off."""
+        """Turn the entity off."""
         self.device.turn_off()
+        self.schedule_update_ha_state()
```
Shouldn't schedule_update_ha_state be only called if turn_off succeeds?

cdnninja replied:
Some other spots have it this way - my theory was after touching devices we should check the source data regardless to ensure we are updated.   Both approaches though in most cases would show correctly I would think.  Which do you prefer?

iprak replied:
The turn_on/off return true if operation succeeds and I think that should be respected and used as the basis of state refresh.

iprak replied:
[Switch ](https://github.com/webdjoe/pyvesync/blob/68abac78f9dc311ba6c7d4077ab2ed75ddd54e28/src/pyvesync/vesyncswitch.py#L174 )returns true if the web call succeeded, we should use that return value to schedule_update_ha_state.

cdnninja replied:
Adjusted.

iprak commented on homeassistant/components/vesync/switch.py:
```diff
@@ -45,53 +71,45 @@ def _setup_entities(
-        self.switch = switch
+    def turn_on(self, **kwargs: Any) -> None:
+        """Turn the entity on."""
+        self.device.turn_on()
```
Shouldn't schedule_update_ha_state be only called if turn_on succeeds?

iprak reviewed (changes requested):
I added some code comments which I encountered while trying to use your branch.

I have added humidifier related switches on top of your branch with some adjustments. You can take a look at my branch here- https://github.com/iprak/core-fork/tree/added-switch-for-humidifier. I will submit a PR for that separately once yours is approved.

cdnninja committed:
feedback elif

cdnninja committed:
Feedback

iprak commented:
> > I added some code comments which I encountered while trying to use your branch.
> > I have added humidifier related switches on top of your branch with some adjustments. You can take a look at my branch here- [iprak/core-fork@`added-switch-for-humidifier`](https://github.com/iprak/core-fork/tree/added-switch-for-humidifier). I will submit a PR for that separately once yours is approved.
> 
> Hopefully it is okay - I factored in some of your items for on_fn and off_fn into this one so it is ready to add future switches easily.

I think I might need to rework this for humidifier. The pyvesync methods for humidifier related switches do not update the internal device state. It is inconsistent, some do and some don't. So I might have to force a refresh instead of state update.

cdnninja commented:
> > > I added some code comments which I encountered while trying to use your branch.
> > > I have added humidifier related switches on top of your branch with some adjustments. You can take a look at my branch here- [iprak/core-fork@`added-switch-for-humidifier`](https://github.com/iprak/core-fork/tree/added-switch-for-humidifier). I will submit a PR for that separately once yours is approved.
> > 
> > 
> > Hopefully it is okay - I factored in some of your items for on_fn and off_fn into this one so it is ready to add future switches easily.
> 
> I think I might need to rework this for humidifier. The pyvesync methods for humidifier related switches do not update the internal device state. It is inconsistent, some do and some don't. So I might have to force a refresh instead of state update.

Okay - I suspect we will have a few things like that as we work through adding items to it.  Hopefully the base sets us up for that.

cdnninja committed:
Add unique ID

cdnninja commented:
Unique IDs added, tests will fail until migration code is done as I think the tests need to exist in that.  Once this branch is updated with the migration and tests PR it should pass.

cdnninja commented:
Unique ID migration PR is ready for review.   Once that is merged I will update this branch and it will also be ready for review.

cdnninja committed:
Merge branch 'dev' into sensorRefactor

iprak commented on homeassistant/components/vesync/const.py:
```diff
@@ -32,6 +37,9 @@
 """Humidifier device types"""
 
+VeSyncFanDevice = VeSyncAirBypass | VeSyncAir131
+"""Humidifier device types"""
```
This is incorrect.

cdnninja replied:
Removed. I feel like that was for fan refactor and had comment errors.

cdnninja committed:
If for on / off

iprak commented:
Any updates on this? I would like to add the last piece for humidifier which is the switch to control display.

cdnninja committed:
remove mistake

cdnninja committed:
Ruff

cdnninja commented:
Cleaned up that last item. I believe ready for merge.  @joostlek
